content/blog/package-react-components/index.md

```diff
@@ -0,0 +1,776 @@
+your new package **Passively compiled**. Which locks you down into this paradigm
+and all its flaws.
+
+I have been able to partially solve this situation by making those
+**Passively compiled** dependencies be `peerDependencies`, but you still need to
+handle them somehow in a test environment (probably by mocking them entirely,
+[Jest][17] does make this fairly easy)
+and if you have any sort of playground or documentation page where
+you display your component then BINGO again, you need to compile
+**Passively compiled** dependencies to make it work.
+
+If this seems crazy to you I can already tell you that I have
+experienced this in my day to day Job and is super frustrating.
+
+#### As Component Consumers
+
+##### Advantages
+
+It is fairly easy to setup with some tweaks to the inclusion paths in your Webpack config.
+
+**Theming** is easy because you just need to overwrite some SASS or LESS variables and your
+build system will do the rest.
+
+##### Disadvantages
+
+You cannot use Webpack abstractions such as [Create React App][16]
+because you always need to customize the sources you include,
+and since this is a rather rare use case,
+most abstractions won't and probably should not provide this functionality.
+
+### Active Compilation
+
+#### As Component Authors
+
+You actually need to write a full web pack configuration
+to generate the bundle, but it is really not that bad (more on this later).
+
+Alternatively, if you are OK with Not-Inlining assets then
+you can use already made solutions such as [create-react-library][105] et al.
+
+You will need to make decisions about how to package and distribute your
+code which _is_ exactly the main point of this post.
+
+#### As Component Consumers
+
+You don't need to customize your Webpack config to use the library,
+this results in simpler configuration. You can also use standard
+pre-made solutions such as [create-react-app][16]
+
+**Theming** will depend exclusively on what the _Component Author_
+sets up to be the Theming API. If a Component Author use
+something like **CSS-in-JS** then the experience will be pretty good.
+[Material-UI React][7] uses this API.
+
+If the Component Author provides a build to compile custom Themes
+out of band (outside your Apps build system) then it will be less
+optimal in my opinion.
+[Ant Design][8] and [React Bootstrap][9] use this API.
+
+## Theming
+
+For Component Libraries or UI Frameworks **Theming** is a _must_. They need to provide a way
+for Component Consumers to change the default look of the components to fit
+their brand or their business.
+
+Lets inspect different strategies of providing Theming and Customization hooks.
+
+### Regular CSS Class Names
+
+This approach works great for single or few component libraries but falls
+very short for larger Component Libraries or UI Frameworks.
+
+The basic premise is that Consumers use well defined CSS class names
+to alter the Component looks.
+
+This approach works only at **Run-Time**.
```

Comment:
What does it really mean that it works only at runtime and not at build time?

Reply:
never mind, I can see the distinction of this versus using CSS/LESS/SASS variables which have to be at build-time (after reading the following sections).

Reply:
I though about having an appendinx or some place else where I talk about build time vs run time, do you think I should reconsider adding it?